### ISSUE TITLE
Multi-User Photo Gallery with private and shared photos/albums

### ISSUE THREAD

**Issue by rkkoszewski** [photoprism/photoprism#98]
It would be great to be able to share a PhotoPrism server between several people, each having their own private pictures/albums, and pictures/albums that could be shared with other users on the same server.

This would probably increase the complexity of the project quite a bit as a whole session/login/profile system would need to be implemented and also handle how to manage the pictures (All pictures in a single folder and assigned to different users by Database, a special folder structure, etc), most likely not compatible with how it works now. 

I was not able to find out from the Wiki about what kind depth is PhotoPrism aiming at regarding photo management. Any input on this from the Devs?

**Comment by lastzero** [comment]
@rkkoszewski This is something we want to do, but also not for the start.

**Comment by lastzero** [comment]
Related [Gitter](https://gitter.im/browseyourlife/community) chat log for future reference:

Michael Mayer @lastzero 14:26
For multi user, we basically just need to add a users table to the db... what gets tricky are ACLs, roles and layered permissions - which is what people often mean with "multiple users". This won't be a free enterprise solution.

Ferdinand Mütsch @n1try 14:26
What I primarily meant by sharing is to have photo albums that are owned by multiple users (friends, family, ...) and are read + write for all of them. Probably a bigger feature

Michael Mayer @lastzero 14:27
I've done all of this for Deutsche Telekom ("Mediencenter") a couple of years ago, so I know what's needed.

Ferdinand Mütsch @n1try 14:29
Yep and user registration logic would probably be handy as well. I get your concerns regarding the free enterprise solution problem

Michael Mayer @lastzero 14:30
I'm all for making progress and adding cool features, especially sharing will be very important when the foundation is implemented in good quality.

Ferdinand Mütsch @n1try 14:31
May I open GitHub feature tickets for multi-user support and shared albums?

Michael Mayer @lastzero 14:31
There should be tickets already
As mentioned, just adding multiple logins is easy but if you don't do it right it can easily end up being a performance and security nightmare
The database should be able to hold the index in memory which is tricky with many users and much more photos, especially when you start to use sharding. This is not done in one or two days like the login itself.

Ferdinand Mütsch @n1try 14:37
I agree...

Michael Mayer @lastzero 14:39
So for a family with not as many photos and no security concerns, it can be done later after thoughtfully thinking about the tradeoffs and implications. This needs to be documented very well too with big warning signs.
However, it would be better for performance to simply run multiple instances and use P2P sharing for albums.
For this, we might provide a central identity service with unique user name.

**Comment by renothing** [comment]
maybe  ActivityPub support is better than multi-user.

**Comment by lastzero** [comment]
@renothing ActivityPub support looks like a great feature for a later release 👍 

Anyone experience with https://github.com/pixelfed/pixelfed?

**Comment by mcombalia** [comment]
I agree just being able to provide multiple login accounts (with the same roles etc) is already a good step forward.

**Comment by muety** [comment]
Any news on this? 🙂

**Comment by oziee** [comment]
i'd like to one day see something like google photos where you can link an account with user/pass to view only selected albums they are assigned to

**Comment by lastzero** [comment]
Working on link sharing next, except you don't create a new account every time you share. You can set a password and expiry date though.

**Comment by oziee** [comment]
is that a per photo things??
i like to add photos to an album and that album just gets bigger and bigger for the grand parents to view photos that get added to it.. just a nice simple "old person" proof way of letting them view all photos in the album

**Comment by lastzero** [comment]
We start with albums, but technically you can share everything incl albums, labels, photos and files.

**Comment by fcimeson** [comment]
Cloud services such as next cloud have already solved this problem. On the server the files are all owned by the cloud server and it decides who can view and edit the files.

**Comment by lastzero** [comment]
You can already upload to and sync with Nextcloud, however image quality will not be the same and users won't see album title, description and other metadata.

**Comment by fcimeson** [comment]
Right, sorry, I wasn't implying you could use nextcloud to solve this problem, I was trying to imply we could use nextcloud solution as a template for solving this problem.

**Comment by ibotty** [comment]
What does "unfounded" in this issue mean?  Can one donate some amount of money specifically to move that feature forward?  I would certainly pay for (a subset of) that feature to appear in the open source version.

**Comment by lastzero** [comment]
See https://docs.photoprism.org/funding/

Yes, you can sponsor this an other issues. Ideally via GitHub Sponsors.

**Comment by anthosz** [comment]
Hi @lastzero,

Do you have more information how it works? The doc is not clear on the priority of issues/improvement.

I mean, the is a goal/deadline (like Wikipédia by example/roadmap to achieve a specific issue? (only for information)

**Comment by lastzero** [comment]
See public roadmap. It's not possible to give you a binding deadline like in a commercial software project unless you agree to cover our full expenses.

**Comment by benmccann** [comment]
One thing I wonder about is how this would work with regards to the file structure on disk. E.g. is there only a single "originals" directory or one per user? If two users import the same photo do we store duplicate copies?

I think I'd lean towards a single deduplicated directory, but I'm not sure I've thought through all the various use cases.

**Comment by lastzero** [comment]
We'll add settings to limit users to sub directories, depending on the use case.

**Comment by benmccann** [comment]
An import directory per user could make sense - then you'd have a way to know which user imported a photo

**Comment by hcsaustrup** [comment]
When implementing this, please support authenticated user headers, so people can use third party authenticating proxies like OAuth2-Proxy :-)

**Comment by mtnbiker777** [comment]
I'd love to be able to give people read only accounts of their own ...

**Comment by txchen** [comment]
Per user per dir might be better, just like DSM moments. It would be easy to import photo to photoprism user by user from other system.
This is now the only road blocker for me to migrate from google photo and moments to photoprism.

**Comment by spoorun** [comment]
It would be great to have 'Group' spaces too - i.e. our community group has a photoprism instance, all photos in that space are shared with all users in the group.

**Comment by michelde** [comment]
> It would be great to have 'Group' spaces too - i.e. our community group has a photoprism instance, all photos in that space are shared with all users in the group.

This goes into the direction as it's been implemented with the new Synology Photos Solution coming next release (DSM 7.0). I already using the Beta Version and very satisfied they merged there old Photostation and Moments Apps into one new app called Photos. There I have a personal photos space (images stored in my personal user space) and there is a public spcae where I can put images all users should have access to. Also I can create albums and share them with different users, which might go into the direction of grouping. Maybe it makes sense to have a look at the ideas of Synology Photo to reuse some of their concepts (if applicable).

**Comment by svengreb** [comment]
Regarding the concerns about the complexity to introduce a user management system, I'd recommend to not start from scratch. One way would be to make use of solutions like [casbin][] for ACL, RBAC and/or ABAC based policies. The project has first-class support for REST APIs and is used in many other large projects.

When it comes to the layout of the files on disk, I would not recommend to reflect the data to the representation of the application. It works fine for the initial indexing and import of new data, but storing files per user/group will add way more complexity and the chance for duplicated data.
The application should map data internally to users/groups while the actual storage path of a file is irrelevant and only required to serve the resource for API endpoints.

[casbin]: https://github.com/casbin/casbin

**Comment by mirisbowring** [comment]
@svengreb are you sure, that casbin would be the way to go in context of access mangement for the files?

It sound great for common roles like admin, member, etc (roles that are able to see specific components / reach specific API Endpoints).

For the files it would be much easier and probably faster to create a mapping table for users who have access to which image (performance would increase in a document based db).

e.x.
Table Users
| ID | Username | ... |
|---|---|---|
| 1 | Bob | ... |
| 2 | Alice | ... |

Table Media
| ID | Hash | ... |
|---|---|---|
| 1 | 7degh6... | ... |
| 2 | h9jzwd... | ... |
| 3 | z4fsjpc... | ... |

Mapping Table
| ID_USER | ID_MEDIA |
|---|---|
| 1 | 2 |
| 1 | 3 |
| 2 | 1 |
| 2 | 2 |
| 2 | 3 |

If i request `/media/7degh6...` , the query would look like this:
```sql
SELECT * FROM MEDIA WHERE ID IN (SELECT ID_MEDIA FROM Mapping_Table WHERE ID_USER == current-user-from-session)
```
Finally, Alice would be able to see the image while Bob receives an `403 Forbidden`

For a more futureproof approach, one could use a `ID_USERGROUP` Map instead of `ID_USER` in case that some images should be shared with a specific group (e.x. the family). If one wants to share the image with a single person, there must be created an internal group that constists of the two parties.

**Comment by FreshMosh** [comment]
I'd also like to fund this as this is the main point that still blocks me from using this as a drop in replacement for Google Photos.

**Comment by muety** [comment]
> I'd also like to fund this as this is the main point that still blocks me from using this as a drop in replacement for Google Photos.

Same for me!

**Comment by Ekleog** [comment]
Curious about this, as it is basically the thing currently delaying me in trying to deploy photoprism: does the plan here include having a tool to merge databases from multiple photoprism instances? In other words, can I currently deploy multiple photoprism instances, one per user, and then when this feature (and esp. shared albums) will land, expect to be able to merge all the databases to make sharing possible later on?

Also, as I see talk of a single deduplicated photo directory. Has the use case “each user has an associated nextcloud user with their own photos directory that is set as the photoprism directory” been taken into consideration, to avoid duplicating all the photos with one in nextcloud and one in photoprism? (though I guess for “group albums” that are rw by multiple people, it would make sense to have a separate photoprism folder as no single user should probably be responsible for the life of the file)

**Comment by lastzero** [comment]
Merging databases would be more like a second or third step and this can be super complex, depending on your requirements. Seems best to keep separate databases / indexed for users who shouldn't see each others content in any way.

**Comment by Qix-** [comment]
# ⚠️  IssueHunt is abandoned.

If you are putting money into it, **you are losing it**. Nobody will be able to receive that money as IssueHunt is not responding to withdrawal requests.

Just a heads up. It's best to just forget about it.

**Comment by lastzero** [comment]
That'd be sad. Where do you have this information from?

**Comment by Qix-** [comment]
People on spectrum, other users, and myself have not heard from them in over a year. They've walked away with the money, it seems. I'm writing Github about it now.

**Comment by lastzero** [comment]
We received money once, but that was some time last year...

**Comment by gt2416** [comment]
Has there been any progress on this feature ? Its really the only thing thats holding me back from using this otherwise grwat product. I think this is a essential feature.

**Comment by lastzero** [comment]
@moximoti is working on it. Sadly, funding was holding us back in the past - and now it looks like IssueHunt is abandoned, so we might never get that money either.

**Comment by Redsandro** [comment]
I'm curious to hear if any of the people that require multi-user or multi-private features for their use-case have used a different privately hosted photo app in the interim. I'm looking for something - perhaps inferior - that can be privately used with mixed sets of users for a temporary project.

Ideally it works on folders so that it can be used in tandem with PhotoPrism, or as a temporary replacement

**Comment by gt2416** [comment]
Yes, I currently use piwigo. It’s nice but doesn’t have object detection. I can’t use photoprism without multi user support. So I’ll stick to piwigo till it comes out at least.

**Comment by jfenton** [comment]
The new Synology Photos is working well for me in lieu of Photoprism having this feature (but it means you need to buy a Synology NAS which is expensive).

J

**Comment by anthosz** [comment]
> 
> The new Synology Photos is working well for me in lieu of Photoprism having this feature (but it means you need to buy a Synology NAS which is expensive). J
> […](#)
> On 19 Jul 2021, at 08:33, gt2416 ***@***.***> wrote: ﻿ Yes, I currently use piwigo. It’s nice but doesn’t have object detection. I can’t use photoprism without multi user support. So I’ll stick to piwigo till it comes out at least. — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub, or unsubscribe.

I use it currently but compared to photoprism, the UI is not good...

**Comment by DanBenHa** [comment]
> [@danbenha](https://issuehunt.io/u/danbenha) has funded $50.00 to this issue.
> 
>     * Submit pull request via [IssueHunt](https://issuehunt.io/repos/119160553/issues/98) to receive this reward.
> 
>     * Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/119160553/issues/98).
> 
>     * Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
> 
>     * Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

Ooops, should have read the whole discussion... Trying to cancel this, since the transaction is fresh 😅 

Would love to push this issue specifically, instead of "just" funding the entire project via Github Sponsors. Maybe in the meantime until an alternative is found, could the number of Github Sponsors that back specific issues be counted?

**Comment by lastzero** [comment]
We count the number of sponsors internally, and also maintain a public roadmap with upcoming features sorted by priority. As written above, @moximoti started working on this issue. It's important to properly implement security sensitive features like this, so development takes a significant amount of time that is well invested. 

Keep in mind: Only getting paid for new features may quickly lead to bad software quality as existing functionality and infrastructure won't be maintained anymore.

We've implemented more features and improvements in the past months than most teams get done in a year, see release notes. That's why we sometimes need a break like everyone else, especially when it's 40°C in our office.

**Comment by DanBenHa** [comment]
> We count the number of sponsors internally, and also maintain a public roadmap with upcoming features sorted by priority. As written above, @moximoti started working on this issue. It's important to properly implement security sensitive features like this, so development takes a significant amount of time that is well invested.
> 
> Keep in mind: Only getting paid for new features may quickly lead to bad software quality as existing functionality and infrastructure won't be maintained anymore.
> 
> We've implemented more features and improvements in the past months than most teams get done in a year, see release notes. That's why we sometimes need a break like everyone else, especially when it's 40°C in our office.

That's good to know. Your point about software quality is well taken. I'm happy to start supporting the project via GH Sponsors (just need to make sure that this issuehunt thing is getting cancelled). 
Nevertheless, this is literally the only feature that's holding me back from using PhotoPrism, so I wanted to give it a little push (recently trying to get my SO to upload vacation pictures to a shared folder on my Nextcloud was not very smooth in comparison to Google Photos 😅).

Wishing you a nice break, if you are having one, or an AC/fan if currently in your office 😄.

(Feel free to mark off-topic. This message doesn't add value to the discussion.)

**Comment by Redsandro** [comment]
> Keep in mind: Only getting paid for new features may quickly lead to bad software quality

Yes, this is a common problem, not new. I'm surprised that IssueHunt/BountySource type solutions don't offer core maintenance fee options.

Ideally, the project owner can set a percentage - e.g. 30% of bounty payouts - that goes to the core team for maintenance of the general project. This way app maintenance is funded by people that rely on the app so much that they are willing to pay for features.

30% seems reasonable. That's what Apple asks for mere app store hosting, and they don't even maintain code to those apps. IssueHunt already takes a whopping 17%. BountySource takes 10%. Core dev team should get at least double of what the bounty payment provider gets imo.

**Comment by DanBenHa** [comment]
Potentially relevant for recent funders via Issuehunt:

I filed a complaint about the credit card transaction with my bank, explaining the intended use and the abandoned state of Issuehunt. Also wrote an email to Issuehunt but, as expected, didn't receive a reply. Got a full refund from my bank (don't know if they somehow cancelled it or if this is an act of goodwill). I'm happy that I can now give this money to the project's team via Github Sponsors :smiley: 

So, if your credit card was charged recently, this might be worth a try.

**Comment by ark-** [comment]
> 
> # ⚠️ IssueHunt is abandoned.
> 
> If you are putting money into it, **you are losing it**. Nobody will be able to receive that money as IssueHunt is not responding to withdrawal requests.
> 
> Just a heads up. It's best to just forget about it.

Should people be cancelling their payments via issuehunt then and switching to, e.g. github sponsors?

**Comment by graciousgrey** [comment]
@ark- we did not try to withdraw money from them recently (as our funded tasks are still in progress).  I have found no official statement about it being abandoned. But some people seem to have problems getting their money.

Anyway we prefer Github Sponsors or Patreon. 
So if you can get a refund from them we are happy to welcome you on Github Sponsors or Patreon :)

**Comment by janstadt** [comment]
Very excited about this feature (hopefully i can convince my wife to use photoprism instead of google photos). Any update on the timeline for this getting into preview?

**Comment by moximoti** [comment]
Hi @janstadt, it's already possible to create additional users via cli in the preview. However, there is no UI and no permission management yet. We're going to work on that after external authentication via OpenID Connect is implemented, which I'm currently working on. There's no exact time schedule, but we hope we'll be able to provide a basic implementation soon.

**Comment by AquaWolf** [comment]
Can you tell us the commands? Will it be persisted when running in a docker container?

**Comment by moximoti** [comment]
Just use `photoprism users --help` to find out about the subcommands. 

E.g. to add a new user, you can just type `photoprism users add`. The cli will then prompt for your infos. To use this command without prompt, just pass username, email, password via flags as indicated in the subcommand's help page (`photoprism users add --help`)

Note, that the `users` subcommands are not final yet and syntax/behavior may still be changed in future iterations.

@AquaWolf yes, the users will be persisted In the database.

**Comment by janstadt** [comment]
> Hi @janstadt, it's already possible to create additional users via cli in the preview. However, there is no UI and no permission management yet. We're going to work on that after external authentication via OpenID Connect is implemented, which I'm currently working on. There's no exact time schedule, but we hope we'll be able to provide a basic implementation soon.

Awesome! Thanks will test it out.

**Comment by Sleuth56** [comment]
Has the users subcommand changed? I'm running version `210523-b1856b9d-Linux-x86_64` in docker and I get `No help topic for 'users'` when I run `photoprism users add` in the container.

**Comment by janstadt** [comment]
> Has the users subcommand changed? I'm running version `210523-b1856b9d-Linux-x86_64` in docker and I get `No help topic for 'users'` when I run `photoprism users add` in the container.

You'll need to pull the latest preview version i believe. Looks like youre running something from May.

**Comment by Sleuth56** [comment]
@janstadt That's exactly the problem. That didn't even cross my mind given that I just pulled the container this morning. Changed to the preview docker tag and it works just fine. Thank you!

**Comment by janstadt** [comment]
Is there a way to limit a specific users permissions for viewing specific originals yet? I've been able to successfully create an additional user but im curious if there is a way to allow them to only be able to access specific folders.

**Comment by Jc2k** [comment]
My understanding of this:

https://github.com/photoprism/photoprism/issues/98#issuecomment-920740596

is no. not yet.

**Comment by janstadt** [comment]
Thanks for pointing that out. I had just gotten to that comment. Lots of action on this issue! Pumped for this feature.

**Comment by guysoft** [comment]
@moximoti Where are the docs on how to get the preview working?

**Comment by lastzero** [comment]
We'll add docs to our User Guide once development is done. Multi user support is not officially released yet. The CLI commands included in our latest release should work though to create and delete (additional) admin users.

**Comment by guysoft** [comment]
I am asking how to set up the preview environment so I can test too and use it.
I can run the preview version behind Basic access authentication. Then security-wise it should not be an issue.

**Comment by lastzero** [comment]
You just need to change the Docker tag from latest to preview, or do you mean develop?

Preview and latest are the same right now as we have just released.

**Comment by hanserasmus** [comment]
Thank you for this product, and this feature! 

Am I correct in deducing so far that the multi-user functionality so far only allows a second "admin" user to be created, and therefor this user will have full access to all files uploaded as well as server settings? I know this is WIP, and it is really great, just want to make sure what I have tested so far does indeed align with what is being developed, and that I am not missing something.

**Comment by Jc2k** [comment]
This comment may help:

https://github.com/photoprism/photoprism/issues/98#issuecomment-920740596

**Comment by hanserasmus** [comment]
@Jc2k Thank you. I have indeed managed to create a second user via the CLI. So my findings are indeed aligned with what is currently being developed. Thank you again!

**Comment by spupuz** [comment]
Thinking about multiuser in case of duplicated photo will be deduplicated and referencend only once?

**Comment by lastzero** [comment]
Right, deduplication needs special attention. We think about separate databases in this case. You see why it takes so long! 😂

**Comment by masterwishx** [comment]
Also waiting for multiuser future :) ,will be great for famaly use...
Using Photoprism Docker in unraid (Linux x64)

**Comment by masterwishx** [comment]
So can we use mariadb for changing some users info after adding users ,for now, befor this future added or better to wait fo version with multiusers ?

**Comment by Jc2k** [comment]
Yes, this feature is still under construction so for your needs it's probably best to wait. I think this comment is still an accurate reflection of the current state of play:

https://github.com/photoprism/photoprism/issues/98#issuecomment-920740596

And here points out that the feature is not released yet:

https://github.com/photoprism/photoprism/issues/98#issuecomment-927295959

Multi-user is a hard thing to do well and can massively slow down database queries if implemented in a rushed way. So the team is rightly taking the time to do it right.

**Comment by Ocean572** [comment]
Hey! Do you all have a timeline on this project?

**Comment by lastzero** [comment]
You mean a deadline?

**Comment by Ocean572** [comment]
@lastzero yes, deadline would be the appropriate comment. Very excited about the feature. Been checking it for months.

**Comment by lastzero** [comment]
We have a zero bug policy and do our best to help users when they need support or have other questions, like right now.

This comes at a price, as we can't give exact deadlines for new features. Also, multi-user is really an epic. Some of it has already been built (run "photorism users help" in a terminal), some of it may be merged into develop this month, and there will be a lot of related features, like sharing, coming sometime next year.

Having said that, funding really has the highest impact. So users can do their part and [become a sponsor](https://docs.photoprism.org/funding/) to get their favorite features as soon as possible.

**Comment by lastzero** [comment]
As you can see in our [Release Notes](https://docs.photoprism.app/release-notes/) and on the public [Roadmap](https://github.com/photoprism/photoprism/projects/5), we've decided not to release additional multi-user features before Christmas:

- This is partly because unexpected [security vulnerabilities](https://docs.photoprism.app/release-notes/) in Go and Log4j surfaced this month and we have a zero bug policy.
- In addition to providing support and fixing bugs discovered in the stable release, multi-user support is currently a top priority.
- Features are not released until we deem them safe and bug free. This is in the best interest of all users, part of our culture, and can't be debated.
- Users can do their part and [become a sponsor](https://docs.photoprism.app/funding/) to get their favorite features as soon as possible. At least until the funding goal is reached, please do not ask for deadlines.

**Comment by IssueHuntBot** [comment]
[@budokaiman](https://issuehunt.io/u/budokaiman) has funded $10.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/119160553/issues/98) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/119160553/issues/98).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by IssueHuntBot** [comment]
[@nylocx](https://issuehunt.io/u/nylocx) has funded $80.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/119160553/issues/98) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/119160553/issues/98).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by graciousgrey** [comment]
Please don't add funds to IssueHunt anymore! 🎁

While we like IssueHunt and are grateful for the donations we've received so far, it hasn't proven to be a sustainable funding option for us as we spend much of our time maintaining existing features and providing support.

If we don't have enough resources to provide support and bugfixes, we can't start working on new features.

**Comment by nylocx** [comment]
Yeah I know but I had $80 more dollars laying around there and as I will not get it out of issue hunt to put it in sponsoring I thought I may just add it to this feature in the hope of a wonder that issue hunt will pay out on this one.

**Comment by IssueHuntBot** [comment]
[@budokaiman](https://issuehunt.io/u/budokaiman) has cancelled funding for this issue.(Cancelled amount: $10.00) [See it on IssueHunt](https://issuehunt.io/repos/119160553/issues/98)

**Comment by wimmme** [comment]
Just a question: would it be possible, or, has anyone tried (if it's possbile) to run a asecondary instance or container for PhotoPrism as a readonly instance ? So running full version where i can manage uploads, tag people and update things, and another instance using the same cache and database in a readonly-mode that i can publish for my family members without them being able to screw things up .... Thats not true Multi-user but might be a good workarround in the mean time ...

**Comment by domsom** [comment]
> Just a question: would it be possible, or, has anyone tried (if it's possbile) to run a asecondary instance or container for PhotoPrism as a readonly instance ?

Tried that by accident: It won't start and log that the storage directory is not writable.

**Comment by lastzero** [comment]
Enabling [read-only mode](https://docs.photoprism.app/getting-started/docker-compose/#volumes) disables all features that require write permission to the *originals* folder, in particular import, upload, and delete. Editing of albums or other metadata stored in the database won't be disabled. The *storage* folder must be writable in any case. You're welcome to share related ideas and howtos in [GitHub Discussions](https://github.com/photoprism/photoprism/discussions) 🗯

Our team will continue to work on multi-user features shortly and release them once they are stable and tested:[^1]

1. We [received a lot of support requests and other questions](https://www.reddit.com/r/photoprism/comments/s26o55/since_we_receive_many_more_support_requests_than/) over the holidays and in the last few days. More than our team can handle.
2. As a result, our priority was to expand the [Troubleshooting Guide](https://docs.photoprism.app/getting-started/troubleshooting/) and divide it into subsections to make it easier for users to find common issues and fix them on their own. Many other parts of our docs were improved as well, for example [Connecting via WebDAV](https://docs.photoprism.app/user-guide/sync/webdav/).
3. We felt that this made more sense than adding more and more features as quickly as possible.

[^1]: Users can do their part and [become a sponsor](https://docs.photoprism.app/funding/) to get their favorite features as soon as possible

**Comment by IssueHuntBot** [comment]
An anonymous user has funded $20.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/119160553/issues/98) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/119160553/issues/98).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by graciousgrey** [comment]
Please don't add funds to IssueHunt anymore! 🎁

While we like IssueHunt and are grateful for the donations we've received so far, it hasn't proven to be a sustainable funding option for us as we spend much of our time maintaining existing features and providing support.

If we don't have enough resources to provide support and bugfixes, we can't start working on new features.

**Comment by IssueHuntBot** [comment]
An anonymous user has cancelled funding for this issue. (Cancelled amount: $20.00) [See it on IssueHunt](https://issuehunt.io/repos/119160553/issues/98)

**Comment by HammyHavoc** [comment]
Can't wait for this, I'll start using Photoprism immediately.

**Comment by Kudjo-Grand** [comment]
I'm also VERY much looking forward to this. I'll be watching for the soonest I can try it. Thank you for all you've built here. I'm very pleased with how it's all coming along.

Keep up the GREAT work!

**Comment by brenty411** [comment]
This would be amazing. I am counting the days until this happens!

**Comment by strugee** [comment]
Hey @brenty411 (and others)! Please don't add +1 comments, as they don't contribute anything to the discussion and just generate more email for the maintainers (and those who are subscribed to this issue) to go through. Instead, you can support issues by using GitHub's reactions feature (smiley icon at the top-right of each comment).

**Comment by IssueHuntBot** [comment]
[@mikebousquet](https://issuehunt.io/u/mikebousquet) has funded $10.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/119160553/issues/98) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/119160553/issues/98).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by IssueHuntBot** [comment]
[@mikebousquet](https://issuehunt.io/u/mikebousquet) has cancelled funding for this issue.(Cancelled amount: $10.00) [See it on IssueHunt](https://issuehunt.io/repos/119160553/issues/98)

**Comment by hawkdavis** [comment]
a good first step would be to allow the creation of a read-only user.

**Comment by graciousgrey** [comment]
@hawkdavis This is the first step and already work  in progress.

**Comment by egodigitus** [comment]
> Just a question: would it be possible, or, has anyone tried (if it's possbile) to run a asecondary instance or container for PhotoPrism as a readonly instance ? So running full version where i can manage uploads, tag people and update things, and another instance using the same cache and database in a readonly-mode that i can publish for my family members without them being able to screw things up .... Thats not true Multi-user but might be a good workarround in the mean time ...

For me running multiple instances is a solid workaround. Make sure to use a different port for the new instance. Example from docker-compose.yml - using 2343 instead of default 2342
```services:
  photoprism:
    ports:
      - "2343:2343" # HTTP port (host:container)
    environment:
      PHOTOPRISM_ADMIN_PASSWORD: "insecure"          # !!! PLEASE CHANGE YOUR INITIAL "admin" PASSWORD !!!
      PHOTOPRISM_SITE_URL: "http://localhost:2343/"  # public server URL incl http:// or https:// and /path, :port is optional
      PHOTOPRISM_HTTP_PORT: "2343"
```
Also make sure you are referencing different folders for your volume mounts

**Comment by xinyo** [comment]
any progress so far?

**Comment by refactoriel** [comment]
> any progress so far?

You see the progress in the commits that link to this issue and you will see that it's done when this issue is closed. Asking for progress doesn't speed up the development and can be annoying to the devs and anyone subscribed to this issue. What DOES help, and has been mentioned several times now, is funding the project via one of the many options available.

**Comment by lastzero** [comment]
> any progress so far?

This is the new login screen that was released last week. What do you think of it? Is this going in the right direction?

![Login Screen](https://dl.photoprism.app/img/ui/login-build-220302-0059f429.jpg)

Check the commit logs and release notes for more changes and improvements:

https://docs.photoprism.app/release-notes/

**Comment by ScuttleSE** [comment]
> 
> You see the progress in the commits that link to this issue and you will see that it's done when this issue is closed. Asking for progress doesn't speed up the development and can be annoying to the devs and anyone subscribed to this issue. What DOES help, and has been mentioned several times now, is funding the project via one of the many options available.

A comment like this doesn't help. It comes across as snooty and rude. The fact that the "issue" has been open for over three years now with very little updates from the developers makes the question 100% legitimate.

I have no doubt the developers are doing their best, but considering it is one of, if not _the_, highest funded feature request, it would be nice with an update and _some_ kind of estimated time when the feature might be available.

**Comment by Ocean572** [comment]
I agree @ScuttleSE. I think the developers are doing great and I was about to donate money after asking the same questions. I then received the same response that you did. Guess where my money didn't go?

**Comment by lastzero** [comment]
So you don't want documentation, support, bugfixes, and any of the other improvements we provide on a regular basis? 🤔

**Comment by Jc2k** [comment]
I do! In fact i've just increased my support level because theres nothing I hate more than entitlement in open source projects. It can be quite toxic, and when it is its disgusting. Let's hoping it doesn't turn toxic here everyone.

Thank you for valuing a stable code base and support!

**Comment by ScuttleSE** [comment]
> So you don't want documentation, support, bugfixes, and any of the other improvements we provide on a regular basis? thinking

That's not what I said, and twisting my argument like that is, frankly, unfair.

Just like I said in my previous post, the mulitple user feature is the most requested and funded feature, treating people asking about it like like they are an annoyance and talking down to them is not, at least in my opinion, a good way to build a community.

If a lot of people constantly are asking about the same thing, don't you think that somewhere information is missing?

I have no problem with you completely axing the feature, or postponing it until 2025, that's fine, but please tell me if you are.
I have invested a lot of time into Photoprism. A lot of that is because over 9 months ago I was told "Actually working on this right now. Also building up a team to make progress faster. Give us a few weeks."
The multiple user feature is something I really need, and back then I thought it was just fine to live without it for a little while until it was polished and released is just fine. But after that I've heard nothing....

**Comment by lastzero** [comment]
This is what we have in store right now. Looking at our feature list, that's a lot and mostly free. There's no reason to be disappointed and act like we've failed.

On the other hand, we have not reached our minimum funding goal even after years (!). Under these (repeatedly communicated) conditions, it is not helpful to ask for a deadline. Answering these questions over and over again - as we are doing now - wastes scarce resources that we could use for development.

As you can see in our commits and roadmap, we have not postponed this feature to 2025. However we have other tasks to take care of as well, and our team is too small.

**Comment by ScuttleSE** [comment]
I'm not sure I completely understand your post. 

Is it this feature that is "in store"? If so, frankly, it has been for 3+ years...

As for funding, that is really unfortunate. I have tried most of all the other self-hosted applications out there and so far, Photoprism is really close to the top. 

I can only speak for myself, but when I tested out Photoprism myself, the "multiple user" feature was kind of a must-have, and when I saw that the feature was on the todo-list I was encouraged. Less so when I saw it had been there for years...

**Comment by lastzero** [comment]
So are other feature requests waiting for years. If you need this for your business, the solution is simple: Help us build a larger team!

**Comment by DataHearth** [comment]
@ScuttleSE Can you move this conversation into Github Discussions, please ? It does not help posting here. it creates annoyance for everyone subscribed to this issue.

If you have something to **help** the maintainers and/or **suggestion** and/or **code** to suggest that might help this feature, please do it.

**Comment by lastzero** [comment]
Note that we stopped using IssueHunt last year, so I removed the IssueHunt comments to make this clear for new users and avoid confusion. I hope GitHub didn't send everyone a notification! If it did, sorry about that :)

**Comment by svengreb** [comment]
@lastzero Maybe a solution for the IssueHunt problem is to ban the @IssueHuntBot account from all repositories in the `photoprism` organization and delete all comments it made up to this day. This way new users, who want to sponsor, won't assume that IssueHunt is the way to go but the <kbd>Sponsor</kbd> button at the top. Sorry for the off-topic, jut wanted to throw this in since you mentioned it.

**Comment by lastzero** [comment]
We have already disabled the IssueHunt bot, but this did not remove existing comments/edits. When we are back in Berlin (on vacation until Friday), we will split this issue into smaller items to make tracking the progress easier! 👍

**Comment by classabbyamp** [comment]
Once this feature is implemented, will there be support for migrating/merging instances?

**Comment by lastzero** [comment]
@classabbyamp Yes, we want to make sure this is possible - that's why we don't want to release a half-baked version that causes more problems than it solves in the long run.

More specifically, we are considering offering two versions:

- a standard version, which has one index and admin, plus optionally additional users with read-only access
- a pro version that transparently runs multiple instances on demand, so you can independently index multiple libraries for multiple users (which is already possible if you set this up manually)

This would allow individual instances to be exported/imported and possibly merged, although it will require a bit more work. The prototype we have right now (and the existing multi-user support that has already been unofficially released) only supports a single index.

Note we just came back from vacation and need to take care of our inboxes before we can proceed with development. Once we have all the details and are sure that the solution is stable, secure, and scalable, we will provide a preview for testing! 🌈

**Comment by daxsis** [comment]
I went through the repo but did not find any task list for this problem. I will start with the easy issues for a while, but I wanted to help with this issue (albeit it is really complex issue). I saw some work done in the commits. No task list or separate branch.

Excuse me if I missed it somewhere, just point me in the right direction please :)

**Comment by lastzero** [comment]
To reduce the number of comments on this topic, please contact us via chat, email or use GitHub Discussions if you have any questions :)

Help with answering support questions and maintaining docs generally has the highest impact on our velocity, as it allows me to spend 4-5 hours a day on development instead of just one. That's a 4x difference.

When we have tasks that are well suited for external contributors, we add a "help wanted" label to those issues (of course, you can always ask if it makes sense to help with other tasks as well).

**Comment by lastzero** [comment]
Note that we currently have [26 issues actively looking for contributions](https://github.com/photoprism/photoprism/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22)  with the "Help Wanted" label. So before anyone asks again, "Is there an update on this?", please check how you can help with those. Thank you! <3

**Comment by gyurmogyuri** [comment]
I will come back if have this.

**Comment by insomnia417** [comment]
I believe this is an old problem, but I don't know why the author is unwilling to add multiple users to help users manage their photo albums. I chose photoprism to replace dsphoto, but without multiple users, I had to create conditions for my needs and choose methods such as nextcloud or filerun to manage photo albums. Is there any meaning in its existence?

**Comment by sixstringsg** [comment]
Have you read the rest of the issue? It had absolutely nothing to do with
willingness. They’ve started to implement the feature and haven’t had the
hours to complete it yet.

**Comment by MiltonFriedman33** [comment]
> @classabbyamp [...]
> More specifically, we are considering offering two versions:
> 
>     * a standard version, which has one index and admin, plus optionally additional users with read-only access
> 
>     * a pro version that transparently runs multiple instances on demand, so you can independently index multiple libraries for multiple users (which is already possible if you set this up manually)
> [...]

I think the majority of people in this thread were expecting multiple users' support to be:
1. Free and open source - (*but I do understand your notion of making such an involving project profitable)
2. Able to create, update and delete multiple users in the same instance (by admin user). A single PhotoPrism instance is already quite resource-consuming, launching multiple instances just to handle multiple users seems like a not optimal workaround.  
3. Users would be able to have private and shared folders with write/read access.

*notice these ideas still allow the pro version to have paid multi-instance support

**Comment by willyp713** [comment]
At a minimum if I could just have an admin login and regular login, that would be a huge win. Longer term, I would pay for separate photo libraries, shared folders, admin controls over access to various folders, etc. But just to start I want to give my family members a login where they don't have all the admin functions.

**Comment by lazee486** [comment]
aside from the shared users, acls etc, any possibility of users as in family members? each user would have a seperate folder path/database....as it is i have to spin up a instance for me, another for my wife, another....I understand charging for enterprise features and acknowledge its a lot of work but before enterprise level is the home user homelab user who is looking for a google photos replacement for their family....

**Comment by masterwishx** [comment]
> At a minimum if I could just have an admin login and regular login, that would be a huge win. 

I think you can do it already in console...

**Comment by lazee486** [comment]
aside from the shared users, acls etc, any possibility of users/profiles as in family members? each user would have a seperate folder path/database....as it is i have to spin up a instance for me, another for my wife, another....I understand charging for enterprise features and acknowledge its a lot of work but before enterprise level is the home user homelab user who is looking for a google photos replacement for their family....

**Comment by kvalev** [comment]
@lastzero kudos for the latest commit regarding multi-user support 👏 

Most of us wont be able to go through all 10-15k lines to see what has changed, do you mind providing a short summary? Am I right to assume that we can now create new accounts (although this was possible before), assign them some non-admin role (maybe even something more fine-grained?) and the users will be able to login and only see and do what they are allowed to?

And how would one go about testing the latest changes? I guess this is not something easily done on the demo instance and most of the users will be reluctant to update their local instances to an alpha version.

**Comment by lastzero** [comment]
As you can see from the amount of code, it's a lot of work. Please be patient as it will take some time to get everything tested, documented and released. The names of the database tables will change, so it is not suitable for production use. Thank you! ❤️

**Comment by DevilsDesigns** [comment]
> As you can see from the amount of code, it's a lot of work. Please be patient as it will take some time to get everything tested, documented and released. The names of the database tables will change, so it is not suitable for production use. Thank you! ❤️

do we need to enable developer updates or how do we enable this feature using the alpha

**Comment by lastzero** [comment]
> do we need to enable developer updates or how do we enable this feature using the alpha

There is no documentation yet and answering everyone's questions individually would not be possible. Therefore, we are only doing internal testing at this time and do not provide public builds, although you could of course compile the source yourself. Once we are satisfied, we will start testing with selected sponsors and customers of our upcoming Pro Edition, as they need these features the most.

Note that we've been focused on development so far and haven't had much time for detailed rollout planning. It seems to make sense to combine enhanced multi-user support with Stripe in the stable release, another feature that has been on our to-do list for far too long. That would make it easier to support us, which is important because we receive more feature requests every day, and I definitely ~~don't want to~~ cannot implement them all on my own.

**Comment by DeltaTangoLima** [comment]
> our upcoming Pro Edition

Just wondering if there's more details you can share on this, yet, please? I've donated once, and will continue to do so annually (I support a few tech projects this way). Wondering if such an approach will open me up to these new features or not.

**Comment by lastzero** [comment]
If you believe our new business offering might be right for you, or if you just want to learn about the differences, you can find the information on our website: https://photoprism.app/teams

For home users, the level of multi-user support - such as the number of roles, users, and security features - may be different for Essentials and Plus: https://photoprism.app/editions

Since everyone receives an email notification when you comment under an issue, we kindly ask that you do not discuss general product or membership questions on GitHub Issues. Thank you for this! We will keep you updated in our chat, Twitter, on our website, and in GitHub Discussions :)

**Comment by ScuttleSE** [comment]
> For home users, the level of multi-user support - such as the number of roles, users, and security features - may be different for Essentials and Plus: https://photoprism.app/editions
> 

I can't see any information on those pages about multi-user support. Also, you only mention Essentials and Plus, does this mean that multi-user support will not be in the Community edition?

**Comment by lastzero** [comment]
We are still in the process of developing the feature. As a result, we haven't had time to plan the rollout in detail or update our website in this regard. Please note that GitHub Issues is not a sales or marketing channel. It is therefore the wrong place to ask these questions.

**Comment by lastzero** [comment]
If you take a moment to check out the source code we've pushed to our public repository - on which the Community Edition is based - you'll find that the feature is there.

However, it needs to be configured for the specific use case, so we may not provide this service - and any other assistance needed to properly use it, such as configuring HTTPS and custom security headers - to those who do not support the project in any way, either financially or through contributions. Especially since the complexity involved leads to additional support requests that we have to deal with.

From our [Membership FAQ](https://photoprism.app/membership) right at the top:

> Think of "free software" as in "free speech," not as in "free beer." The [Free Software Foundation](https://www.gnu.org/philosophy/free-sw.en.html) sometimes calls it "libre software," borrowing the French or Spanish word for "free" as in freedom, to show they do not mean the software is gratis.

This has always been our policy, so nothing is changing. As for the details, we ask for a little time to work them out together with our community! ❤️

**Comment by lastzero** [comment]
**You are now invited to help test the latest preview build!** Before you upgrade, be sure to [read the full release notes](https://docs.photoprism.app/release-notes/#development-preview). We appreciate your patience and feedback on the changes.

**Comment by ScuttleSE** [comment]
Hm, am I mistaken, or is there still no way to have multiple users in the preview-build? I can't find any place in the UI where I can create users.

**Comment by lastzero** [comment]
You can now edit your own user profile in the UI and the new implementation is also much more secure. However, there is no web UI for adding users yet, only the newly developed CLI commands - for security reasons I tend not to build a public API for user management, but to find other solutions (which still needs to be done).

If we don't get help "testing" until everything is 100% done and working, that's not very helpful (at least for now, that's why I said users can test later - note that GitHub Issues is primarily for contributors who aren't afraid of using a terminal).

There are many more features, enhancements, and bug fixes in the preview, so it would be great if we could release them before the multi-user feature is 100% complete. The less time we have to spend testing and supporting the next release, the faster we can move forward with adding multi-user functionality.

**Comment by Laptop765** [comment]
Thank you so much for all of this work! For those of us who want to test
multi-user through the CLI is there anything we need to know about future
compatibility, e.g. something is expected to change that might require
starting additional users over? Or are things stable enough within reason
bearing in mind that it’s still in development?

**Comment by lastzero** [comment]
Thanks for asking! For now, our focus is on stability and getting feedback regarding the changes mentioned in the release notes. In particular, the link sharing view is now based on the regular UI and a Visitor role. That's the only new role we currently offer/test in addition to the standard Admin role. If those work and the existing admin account can be properly migrated from the last stable version, we can move forward AFTER an update has been released for those who not exclusively wait for multi user only, but for example have HEIC or ProRAW files that currently don't work as they should. Future developments will be compatible with that stable release, but not necessarily with the current development version in case we need to implement incompatible changes in response to test results.

**Comment by ark-** [comment]
Excellent work :) 

What is the visitor role for?

**Comment by lastzero** [comment]
Link visitors when sharing albums or folders using a secret link. In the future, you will also be able to share single files and more this way.

**Comment by ark-** [comment]
Ah ok, I forced creation of a visitor in the command line using the following and couldn’t log in:

photoprism users add name -r visitor

Wanted to make sure that wasn’t a bug, sounds like it’s not. So the share function makes its own visitor account when creating a passworded link?

**Comment by lastzero** [comment]
@ark- Yes, there is a built-in "Visitor" account with the user ID "-2". You do not need to create it manually. It cannot log in with username and password, but only with share tokens. Previously it had the "Guest" role, which we now want to use for identified guests with passwords. So it is important to test if the existing functionality works as before (or better).

Edit: There should be the option to either set up guest accounts for people you know and want to share an album with (without setting a new password for each album you share), or protected visitor links that require a password to activate but don't create a new user account for each link visitor (which can be many, depending on the use case).

**Comment by suaide** [comment]
> @ark- Yes, there is a built-in "Visitor" account with the user ID "-2". You do not need to create it manually. It cannot log in with username and password, but only with share tokens. Previously it had the "Guest" role, which we now want to use for identified guests with passwords. So it is important to test if the existing functionality works as before (or better).
> 
> Edit: There should be the option to either set up guest accounts for people you know and want to share an album with (without setting a new password for each album you share), or protected visitor links that require a password to activate but don't create a new user account for each link visitor (which can be many, depending on the use case).

Hi, this is a great improvement. I am wondering what are the available roles. I see I can have an 'admin' role and a 'visitor' role. Visitor can not log in with user/pass. Is there any other role available, such as user, etc?

**Comment by silvermd** [comment]
Hello, I'm a silent reader and just want to give my attention here. I also wish mulituser. However, I am currently using truenas scale (app) and have to wait for an implementation and another solution to create the user. Thanks for all the effort, I'm really looking forward to the results.

Thanks, @ all

**Comment by lastzero** [comment]
Thanks for your feedback! We'll keep that in mind and will continue working on this issue after the existing improvements have been released, so users can already enjoy those and don't need to wait much longer.

**Comment by SagarMhatre** [comment]
Is there a way to get the visitor in a docker image ?

**Comment by lastzero** [comment]
Use our preview image and create an album share link you open in a private browser tab.

**Comment by lastzero** [comment]
After today's changes, it looks like we can release the current version as an update. Native speakers are invited to [add missing user interface translations for their language](https://docs.photoprism.app/developer-guide/translations-weblate/). We will wait another day or two for that. Thanks to [everyone who contributed](https://docs.photoprism.app/developer-guide/)!

**Comment by lastzero** [comment]
Be aware that multi-user support will be finalized **after the upcoming release**, as it includes many other improvements not related to user management. We do not want to delay these any further. This is also [mentioned in the release notes](https://docs.photoprism.app/release-notes/), which you should read before upgrading.

**Please don't push forward with multi-user enhancements yet, as this is work in progress and we may not be able to merge your pull request.** If you want to help with something other than [updating the translations](https://docs.photoprism.app/developer-guide/translations-weblate/), feel free to take a look at the [issues labeled "help wanted"](https://github.com/photoprism/photoprism/issues?q=is%3Aissue+is%3Aopen+label%3A%22help+wanted%22).

**Comment by DeltaTangoLima** [comment]
I'm so excited to see some of the progress the team are making. Quick question: for those that moved to the preview release, when would be the right time to move back to latest stable, without losing any of the new functionality we've been trialing?

**Comment by lastzero** [comment]
Depends on which preview release. The current preview is the release candidate, so it could be identical with the upcoming release.

**Comment by SaswatPadhi** [comment]
> @ark- Yes, there is a built-in "Visitor" account with the user ID "-2". You do not need to create it manually. It cannot log in with username and password, but only with share tokens. Previously it had the "Guest" role, which we now want to use for identified guests with passwords. So it is important to test if the existing functionality works as before (or better).

I was playing with `photoprism users` command and I created a new user as `photoprism user add test`.

With `photoprism users ls`, I see that both `test` and `admin` accounts have `admin` role.
However, the "Super Admin" field has `no` for `test` vs `yes` for `admin`.

I tried a quick grep but couldn't find anything much on the "Super Admin" capability. Could you please explain the additional rights granted by `yes` on "Super Admin"? Thanks!

**Comment by graciousgrey** [comment]
The role of a user with superadmin status cannot be changed, whereas a user with the admin role can be assigned a different role. Documentation will follow once we have introduced more multi-user functionality such as additional roles.

**Comment by jojo-gfx** [comment]
Uaah, have seen it now that photoprism have no usermanagement after I setup it with 60k photos and now... uaaah!
This feature had to implement from beginning/planning the software ... urgs! Now it will be a heavy problem, it to implement after it is growing so big.
Hopefully you will have success with it because it is THE feature what is making the software incredible or relative senseless!
Hope!!!!!!!! ... because the generally functions are great! ;)

A suggestion from me, because many many OSS projects implement the usermanagement in the main sowftwar layer and makes the whole system big, slow and in worst case unusable. Please use the tadabase here as slave to do this perfectly database driven work parts! E.g. the Typo3 project never changed their system to database driven management of rights and at time big projects with many users and content are horrible slow because this old programming failure.
Databases are incredible efficient in intersection of massive data ammounts and this is, what usermanagement does! :) So you also can keep your code clean with no trips into the hell!
Also proffesional developers often use databases only as stupid data storage, but they all forget, databases are mega intelligent and absolute the fastest way to use data and its context between.

The most software developers don't know, what a database is... bad thing! Hopefully here it will be better! :)

Generally, I'm a big fan of photoprism so far... :D

*near by:
Ah, by the way... would be nice you cloud create a guest-view, where any content could be generally offered to users as display window.  Also, at time with all album sharing links you can view all other shared albums also. Ao at time is works like, what I mean, but on a currious way. ;) ... you have to autenticate with any sharing link and then you can use it free. 
Would think, it's still a bug and it's also based on not implemented databasedriven usermanagement. :D 

Also would be nice to have here, and also in the the sharing per link, the map available, because it's a killer feature here what is really really great, but only for the adminuser it's a shame because nobody around can use this incredible nice working killerfeature!

... have taken a look in the generated database and looks not so good per first short view.
Seems usermanagement packed into the servers programming, but only per short look at time.
Could it be, one user can only match one group? No good Idea. Please think here about groups who increment features/rights and add them to a combination, what can ad to a user, so not he is only admin or not ... because also an admin has the normal functions of a normal user and admin is only a addon to normal user e.g.. otherwise you will manage this in the program as own cases and will hit you in an heavy ammount of complexity and no flexible structures, also in look of new functions in the future. So far what I've seen now ... don't look, it's true. ;)

**Comment by lastzero** [comment]
We would like to remind you that this is not a forum and that you can use GitHub Discussions for general questions and comments:

- https://github.com/photoprism/photoprism/discussions/1678

Thank you! 💖

**Comment by lastzero** [comment]
I've updated the title of this issue and changed the label to "epic" since initial multi-user support for those [sharing a single library](https://docs.photoprism.app/user-guide/users/libraries/) [was released in 2023](https://docs.photoprism.app/release-notes/#may-2-2023), with [2-Factor Authentication](https://docs.photoprism.app/user-guide/users/2fa/) and [Single Sign-On via OpenID Connect](https://docs.photoprism.app/getting-started/advanced/openid-connect/) added this year:

- https://github.com/photoprism/photoprism/releases/tag/230502-c405f6eff
- https://github.com/photoprism/photoprism/issues/808
- https://github.com/photoprism/photoprism/issues/782

What's still missing to fully satisfy this feature request is the ability to easily set up and index multiple independent libraries on the same instance (so you don't have to set up multiple instances manually), and an improved user interface that makes sharing with other users more convenient and obvious than [sending them links](https://docs.photoprism.app/user-guide/users/sharing/) to click on:

- https://github.com/photoprism/photoprism/issues/557
- you can click the https://github.com/photoprism/photoprism/labels/sharing label to see all sharing-related issues

Note that in order to avoid duplicate work and merge conflicts, we can only start implementing this once the **ongoing upgrade of our frontend component framework** is complete:

- https://github.com/photoprism/photoprism/issues/3168

This also applies to [other frontend-related issues](https://github.com/photoprism/photoprism/labels/frontend) such as [batch editing](https://github.com/photoprism/photoprism/issues/271) and [selecting an album cover image](https://github.com/photoprism/photoprism/issues/383). Thank you very much for your patience and continued support! 🌈💎✨
